File: HistName.cs
// HistName.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace HistName
{
    class Program
    {
        static void Main(string[] args)
        {
            var opts = ParseArgs(args);
            var gen = new HistoricalNameGenerator(opts);
            var names = gen.Generate();
            if (opts.Output != null)
            {
                string fmt = opts.Format;
                if (string.IsNullOrEmpty(fmt))
                {
                    string ext = Path.GetExtension(opts.Output).ToLower().TrimStart('.');
                    fmt = ext == "json" ? "json" : ext == "csv" ? "csv" : "text";
                }
                switch (fmt)
                {
                    case "json": ExportJson(names, opts.Culture, opts.Gender, opts.Output); break;
                    case "csv": ExportCsv(names, opts.Output); break;
                    default: ExportText(names, opts.Output); break;
                }
                Console.WriteLine($"Результат сохранён в {opts.Output}");
            }
            else
            {
                PrintNames(names, opts.Culture, opts.Gender, !Console.IsOutputRedirected);
            }
        }

        static Options ParseArgs(string[] args)
        {
            var opts = new Options();
            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--culture": opts.Culture = args[++i]; break;
                    case "--gender": opts.Gender = args[++i]; break;
                    case "--count": opts.Count = int.Parse(args[++i]); break;
                    case "--surname": opts.Surname = true; break;
                    case "--title": opts.Title = true; break;
                    case "--seed": opts.Seed = long.Parse(args[++i]); break;
                    case "--output": opts.Output = args[++i]; break;
                    case "--format": opts.Format = args[++i]; break;
                }
            }
            return opts;
        }

        class Options
        {
            public string Culture { get; set; } = "viking";
            public string Gender { get; set; } = "male";
            public int Count { get; set; } = 1;
            public bool Surname { get; set; } = false;
            public bool Title { get; set; } = false;
            public long? Seed { get; set; }
            public string Output { get; set; }
            public string Format { get; set; }
        }

        class HistoricalNameGenerator
        {
            private static readonly Dictionary<string, Dictionary<string, List<string>>> Names = new()
            {
                ["viking"] = new()
                {
                    ["male"] = new() { "Эйнар","Рагнар","Бьёрн","Свен","Ульф","Харальд","Олаф","Эрик","Торгейр","Гуннар","Сигурд","Хакон","Кнуд","Орм","Кетиль","Асбьёрн","Торстейн","Эгиль","Гудмунд","Хельги" },
                    ["female"] = new() { "Астрид","Брюнхильд","Гудрун","Сигриди","Гунхильд","Хельга","Ингеборг","Тордис","Тора","Альвильд","Герда","Фрейдис","Раннвейг","Сигрид","Эрна","Грета","Кристин","Хильда","Руна","Сванхильд" },
                    ["surnames"] = new() { "Бьёрнсон","Эйнарсон","Харальдсон","Свенсон","Рагнарсон","Ульфсон","Олафсон","Гуннарсон","Сигурдсон","Торгерсон","Кетильсон","Асбьёрнсон","Торстейнсон","Эгильсон","Гудмундсон" },
                    ["titles"] = new() { "Берсерк","Скальд","Молот","Боевой Топор","Кровавый","Северный","Грозный","Мудрый" }
                },
                ["roman"] = new()
                {
                    ["male"] = new() { "Гай","Луций","Марк","Квинт","Публий","Секст","Тит","Авл","Децим","Гней","Сервий","Нумерий","Август","Тиберий","Клавдий","Нерон","Траян","Адриан","Антоний","Юлий" },
                    ["female"] = new() { "Юлия","Ливия","Валерия","Клавдия","Цецилия","Теренция","Октавия","Друзилла","Агриппина","Мессалина","Антония","Корнелия","Помпея","Вителлия","Германика","Фавстина","Луцилла","Криспина","Сабина","Вероника" },
                    ["surnames"] = new() { "Цезарь","Август","Нерон","Клавдий","Тиберий","Траян","Адриан","Антоний","Юлий","Германик","Брут","Сципион","Катон","Цицерон","Сенека","Аврелий","Флавий","Константин" },
                    ["titles"] = new() { "Император","Консул","Сенатор","Легат","Претор","Эдил","Цензор","Диктатор" }
                },
                ["medieval"] = new()
                {
                    ["male"] = new() { "Уильям","Генри","Ричард","Роберт","Эдуард","Джон","Томас","Джеффри","Роджер","Хью","Саймон","Уолтер","Гилберт","Реджинальд","Балдуин","Алан","Брайан","Филип","Джеймс","Дэвид" },
                    ["female"] = new() { "Элеонора","Матильда","Алиенора","Агнес","Изабелла","Беатрис","Сибилла","Маргарет","Кэтрин","Джоан","Алиса","Сесилия","Ида","Мария","Клеменция","Амелия","Филиппа","Гвендолин","Аделаида","Энн" },
                    ["surnames"] = new() { "Нормандский","Анжуйский","Ланкастерский","Йоркский","Плантагенет","Уэссекский","Глостерский","Кентский","Суффолкский","Ричмондский","Пемброк","Мортимер","Грей","Стюарт","Говард" },
                    ["titles"] = new() { "Король","Герцог","Граф","Барон","Рыцарь","Лорд","Епископ","Маркиз" }
                },
                ["greek"] = new()
                {
                    ["male"] = new() { "Александр","Аристотель","Демокрит","Эпикур","Гераклит","Платон","Сократ","Пифагор","Гомер","Софокл","Еврипид","Эсхил","Перикл","Леонид","Фемистокл","Агамемнон","Ахилл","Одиссей","Гектор","Тесей" },
                    ["female"] = new() { "Афина","Гера","Афродита","Артемида","Деметра","Персефона","Гестия","Ника","Ирида","Афродита","Кассандра","Андромеда","Елена","Пенелопа","Антигона","Электра","Ифигения","Медея","Сапфо","Гипполита" },
                    ["surnames"] = new() { "Афинский","Спартанский","Коринфский","Фиванский","Македонский","Фессалийский","Критский","Родосский","Милетский","Эфесский" },
                    ["titles"] = new() { "Царь","Тиран","Архонт","Стратег","Философ","Поэт","Воин","Герой" }
                },
                ["slavic"] = new()
                {
                    ["male"] = new() { "Владимир","Святослав","Ярослав","Всеволод","Изяслав","Мстислав","Олег","Игорь","Святополк","Вячеслав","Ярополк","Глеб","Борис","Андрей","Александр","Дмитрий","Михаил","Пётр","Иван","Василий" },
                    ["female"] = new() { "Ольга","Ярославна","Владислава","Людмила","Мирослава","Светлана","Евдокия","Борислава","Велеслава","Добромира","Вера","Надежда","Любовь","Мария","Анна","Екатерина","Анастасия","Татьяна","Елена","Ирина" },
                    ["surnames"] = new() { "Рюрикович","Владимирович","Святославич","Ярославич","Всеволодович","Изяславич","Мстиславич","Олегович","Игоревич","Святополкович" },
                    ["titles"] = new() { "Князь","Воевода","Боярин","Дружинник","Мудрый","Грозный","Святой","Великий" }
                },
                ["celtic"] = new()
                {
                    ["male"] = new() { "Айдан","Брендан","Коннор","Деклан","Эйлин","Финн","Гэвин","Иан","Киран","Лайам","Мэлколм","Ниалл","Она","Патрик","Рори","Шон","Тиадг","Улисс","Эмон","Брайан" },
                    ["female"] = new() { "Айслин","Бриджит","Кэтлин","Диана","Эйлиш","Финола","Грэйн","Иона","Кира","Лианна","Майв","Ниам","Она","Рона","Сиана","Тара","Уна","Фиона","Сиобан","Морриган" },
                    ["surnames"] = new() { "МакКауд","О'Брайен","О'Салливан","МакКарти","О'Доннелл","МакДонах","О'Нейл","МакГрат","О'Ши","МакМахон","О'Коннор","МакДауэлл","О'Киф","МакГиннис","О'Флаэрти" },
                    ["titles"] = new() { "Король","Вождь","Друид","Воин","Мудрец","Певец","Охотник","Мастер" }
                }
            };

            private readonly string culture, gender;
            private readonly int count;
            private readonly bool surname, title;
            private readonly Random rng;
            private readonly Dictionary<string, List<string>> cultureData;

            public HistoricalNameGenerator(Options opts)
            {
                culture = opts.Culture;
                gender = opts.Gender;
                count = opts.Count;
                surname = opts.Surname;
                title = opts.Title;
                rng = opts.Seed.HasValue ? new Random((int)opts.Seed.Value) : new Random();
                cultureData = Names.GetValueOrDefault(culture, Names["viking"]);
            }

            public string GenerateName()
            {
                var names = cultureData.GetValueOrDefault(gender, cultureData["male"]);
                string name = names[rng.Next(names.Count)];
                if (surname && cultureData.ContainsKey("surnames"))
                {
                    var snames = cultureData["surnames"];
                    name += " " + snames[rng.Next(snames.Count)];
                }
                if (title && cultureData.ContainsKey("titles"))
                {
                    var titles = cultureData["titles"];
                    name += " " + titles[rng.Next(titles.Count)];
                }
                return name;
            }

            public List<string> Generate()
            {
                var list = new List<string>();
                for (int i = 0; i < count; i++) list.Add(GenerateName());
                return list;
            }
        }

        static void PrintNames(List<string> names, string culture, string gender, bool color)
        {
            var labels = new Dictionary<string, string>
            {
                ["viking"] = "викинг", ["roman"] = "римское", ["medieval"] = "средневековое",
                ["greek"] = "греческое", ["slavic"] = "славянское", ["celtic"] = "кельтское"
            };
            string label = labels.GetValueOrDefault(culture, culture);
            string genderLabel = gender == "male" ? "мужские" : "женские";
            if (color)
            {
                Console.ForegroundColor = ConsoleColor.Cyan;
                Console.WriteLine($"🏛️ Исторические имена ({label}, {genderLabel}):");
                Console.ResetColor();
                for (int i = 0; i < names.Count; i++)
                {
                    Console.ForegroundColor = ConsoleColor.Green;
                    Console.WriteLine($"{i+1}. {names[i]}");
                    Console.ResetColor();
                }
            }
            else
            {
                Console.WriteLine($"Исторические имена ({label}, {genderLabel}):");
                for (int i = 0; i < names.Count; i++) Console.WriteLine($"{i+1}. {names[i]}");
            }
        }

        static void ExportJson(List<string> names, string culture, string gender, string filename)
        {
            var data = new { culture, gender, names };
            string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(filename, json);
        }

        static void ExportCsv(List<string> names, string filename)
        {
            using var sw = new StreamWriter(filename);
            sw.WriteLine("name");
            foreach (var n in names) sw.WriteLine(n);
        }

        static void ExportText(List<string> names, string filename)
        {
            File.WriteAllText(filename, string.Join("\n", names));
        }
    }
}

File: hist_name.rs
// hist_name.rs
use clap::{App, Arg};
use rand::prelude::*;
use rand::SeedableRng;
use rand::rngs::StdRng;
use serde_json;
use std::collections::HashMap;
use std::fs;
use std::io::Write;

lazy_static::lazy_static! {
    static ref NAMES_DATA: HashMap<String, HashMap<String, Vec<String>>> = {
        let mut data = HashMap::new();

        let mut viking = HashMap::new();
        viking.insert("male".to_string(), vec!["Эйнар","Рагнар","Бьёрн","Свен","Ульф","Харальд","Олаф","Эрик","Торгейр","Гуннар","Сигурд","Хакон","Кнуд","Орм","Кетиль","Асбьёрн","Торстейн","Эгиль","Гудмунд","Хельги"].iter().map(|s| s.to_string()).collect());
        viking.insert("female".to_string(), vec!["Астрид","Брюнхильд","Гудрун","Сигриди","Гунхильд","Хельга","Ингеборг","Тордис","Тора","Альвильд","Герда","Фрейдис","Раннвейг","Сигрид","Эрна","Грета","Кристин","Хильда","Руна","Сванхильд"].iter().map(|s| s.to_string()).collect());
        viking.insert("surnames".to_string(), vec!["Бьёрнсон","Эйнарсон","Харальдсон","Свенсон","Рагнарсон","Ульфсон","Олафсон","Гуннарсон","Сигурдсон","Торгерсон","Кетильсон","Асбьёрнсон","Торстейнсон","Эгильсон","Гудмундсон"].iter().map(|s| s.to_string()).collect());
        viking.insert("titles".to_string(), vec!["Берсерк","Скальд","Молот","Боевой Топор","Кровавый","Северный","Грозный","Мудрый"].iter().map(|s| s.to_string()).collect());
        data.insert("viking".to_string(), viking);

        let mut roman = HashMap::new();
        roman.insert("male".to_string(), vec!["Гай","Луций","Марк","Квинт","Публий","Секст","Тит","Авл","Децим","Гней","Сервий","Нумерий","Август","Тиберий","Клавдий","Нерон","Траян","Адриан","Антоний","Юлий"].iter().map(|s| s.to_string()).collect());
        roman.insert("female".to_string(), vec!["Юлия","Ливия","Валерия","Клавдия","Цецилия","Теренция","Октавия","Друзилла","Агриппина","Мессалина","Антония","Корнелия","Помпея","Вителлия","Германика","Фавстина","Луцилла","Криспина","Сабина","Вероника"].iter().map(|s| s.to_string()).collect());
        roman.insert("surnames".to_string(), vec!["Цезарь","Август","Нерон","Клавдий","Тиберий","Траян","Адриан","Антоний","Юлий","Германик","Брут","Сципион","Катон","Цицерон","Сенека","Аврелий","Флавий","Константин"].iter().map(|s| s.to_string()).collect());
        roman.insert("titles".to_string(), vec!["Император","Консул","Сенатор","Легат","Претор","Эдил","Цензор","Диктатор"].iter().map(|s| s.to_string()).collect());
        data.insert("roman".to_string(), roman);

        let mut medieval = HashMap::new();
        medieval.insert("male".to_string(), vec!["Уильям","Генри","Ричард","Роберт","Эдуард","Джон","Томас","Джеффри","Роджер","Хью","Саймон","Уолтер","Гилберт","Реджинальд","Балдуин","Алан","Брайан","Филип","Джеймс","Дэвид"].iter().map(|s| s.to_string()).collect());
        medieval.insert("female".to_string(), vec!["Элеонора","Матильда","Алиенора","Агнес","Изабелла","Беатрис","Сибилла","Маргарет","Кэтрин","Джоан","Алиса","Сесилия","Ида","Мария","Клеменция","Амелия","Филиппа","Гвендолин","Аделаида","Энн"].iter().map(|s| s.to_string()).collect());
        medieval.insert("surnames".to_string(), vec!["Нормандский","Анжуйский","Ланкастерский","Йоркский","Плантагенет","Уэссекский","Глостерский","Кентский","Суффолкский","Ричмондский","Пемброк","Мортимер","Грей","Стюарт","Говард"].iter().map(|s| s.to_string()).collect());
        medieval.insert("titles".to_string(), vec!["Король","Герцог","Граф","Барон","Рыцарь","Лорд","Епископ","Маркиз"].iter().map(|s| s.to_string()).collect());
        data.insert("medieval".to_string(), medieval);

        let mut greek = HashMap::new();
        greek.insert("male".to_string(), vec!["Александр","Аристотель","Демокрит","Эпикур","Гераклит","Платон","Сократ","Пифагор","Гомер","Софокл","Еврипид","Эсхил","Перикл","Леонид","Фемистокл","Агамемнон","Ахилл","Одиссей","Гектор","Тесей"].iter().map(|s| s.to_string()).collect());
        greek.insert("female".to_string(), vec!["Афина","Гера","Афродита","Артемида","Деметра","Персефона","Гестия","Ника","Ирида","Афродита","Кассандра","Андромеда","Елена","Пенелопа","Антигона","Электра","Ифигения","Медея","Сапфо","Гипполита"].iter().map(|s| s.to_string()).collect());
        greek.insert("surnames".to_string(), vec!["Афинский","Спартанский","Коринфский","Фиванский","Македонский","Фессалийский","Критский","Родосский","Милетский","Эфесский"].iter().map(|s| s.to_string()).collect());
        greek.insert("titles".to_string(), vec!["Царь","Тиран","Архонт","Стратег","Философ","Поэт","Воин","Герой"].iter().map(|s| s.to_string()).collect());
        data.insert("greek".to_string(), greek);

        let mut slavic = HashMap::new();
        slavic.insert("male".to_string(), vec!["Владимир","Святослав","Ярослав","Всеволод","Изяслав","Мстислав","Олег","Игорь","Святополк","Вячеслав","Ярополк","Глеб","Борис","Андрей","Александр","Дмитрий","Михаил","Пётр","Иван","Василий"].iter().map(|s| s.to_string()).collect());
        slavic.insert("female".to_string(), vec!["Ольга","Ярославна","Владислава","Людмила","Мирослава","Светлана","Евдокия","Борислава","Велеслава","Добромира","Вера","Надежда","Любовь","Мария","Анна","Екатерина","Анастасия","Татьяна","Елена","Ирина"].iter().map(|s| s.to_string()).collect());
        slavic.insert("surnames".to_string(), vec!["Рюрикович","Владимирович","Святославич","Ярославич","Всеволодович","Изяславич","Мстиславич","Олегович","Игоревич","Святополкович"].iter().map(|s| s.to_string()).collect());
        slavic.insert("titles".to_string(), vec!["Князь","Воевода","Боярин","Дружинник","Мудрый","Грозный","Святой","Великий"].iter().map(|s| s.to_string()).collect());
        data.insert("slavic".to_string(), slavic);

        let mut celtic = HashMap::new();
        celtic.insert("male".to_string(), vec!["Айдан","Брендан","Коннор","Деклан","Эйлин","Финн","Гэвин","Иан","Киран","Лайам","Мэлколм","Ниалл","Она","Патрик","Рори","Шон","Тиадг","Улисс","Эмон","Брайан"].iter().map(|s| s.to_string()).collect());
        celtic.insert("female".to_string(), vec!["Айслин","Бриджит","Кэтлин","Диана","Эйлиш","Финола","Грэйн","Иона","Кира","Лианна","Майв","Ниам","Она","Рона","Сиана","Тара","Уна","Фиона","Сиобан","Морриган"].iter().map(|s| s.to_string()).collect());
        celtic.insert("surnames".to_string(), vec!["МакКауд","О'Брайен","О'Салливан","МакКарти","О'Доннелл","МакДонах","О'Нейл","МакГрат","О'Ши","МакМахон","О'Коннор","МакДауэлл","О'Киф","МакГиннис","О'Флаэрти"].iter().map(|s| s.to_string()).collect());
        celtic.insert("titles".to_string(), vec!["Король","Вождь","Друид","Воин","Мудрец","Певец","Охотник","Мастер"].iter().map(|s| s.to_string()).collect());
        data.insert("celtic".to_string(), celtic);

        data
    };
}

struct Generator {
    culture: String,
    gender: String,
    count: usize,
    surname: bool,
    title: bool,
    rng: StdRng,
}

impl Generator {
    fn new(culture: &str, gender: &str, count: usize, surname: bool, title: bool, seed: u64) -> Self {
        let rng = if seed != 0 { StdRng::seed_from_u64(seed) } else { StdRng::from_entropy() };
        Generator {
            culture: culture.to_string(),
            gender: gender.to_string(),
            count,
            surname,
            title,
            rng,
        }
    }

    fn generate_name(&mut self) -> String {
        let culture_data = NAMES_DATA.get(&self.culture).unwrap_or_else(|| NAMES_DATA.get("viking").unwrap());
        let names = culture_data.get(&self.gender).unwrap_or_else(|| culture_data.get("male").unwrap());
        let mut name = names[self.rng.gen_range(0..names.len())].clone();
        if self.surname {
            if let Some(surnames) = culture_data.get("surnames") {
                name.push(' ');
                name.push_str(&surnames[self.rng.gen_range(0..surnames.len())]);
            }
        }
        if self.title {
            if let Some(titles) = culture_data.get("titles") {
                name.push(' ');
                name.push_str(&titles[self.rng.gen_range(0..titles.len())]);
            }
        }
        name
    }

    fn generate(&mut self) -> Vec<String> {
        (0..self.count).map(|_| self.generate_name()).collect()
    }
}

fn print_names(names: &[String], culture: &str, gender: &str, color: bool) {
    let labels = vec![
        ("viking", "викинг"), ("roman", "римское"), ("medieval", "средневековое"),
        ("greek", "греческое"), ("slavic", "славянское"), ("celtic", "кельтское")
    ];
    let label = labels.iter().find(|(k,_)| *k == culture).map(|(_,v)| *v).unwrap_or(culture);
    let gender_label = if gender == "male" { "мужские" } else { "женские" };
    if color {
        println!("\x1b[36m🏛️ Исторические имена ({}, {}):\x1b[0m", label, gender_label);
        for (i, n) in names.iter().enumerate() {
            println!("{}. \x1b[32m{}\x1b[0m", i+1, n);
        }
    } else {
        println!("Исторические имена ({}, {}):", label, gender_label);
        for (i, n) in names.iter().enumerate() {
            println!("{}. {}", i+1, n);
        }
    }
}

fn export_json(names: &[String], culture: &str, gender: &str, filename: &str) {
    let data = serde_json::json!({
        "culture": culture,
        "gender": gender,
        "names": names
    });
    let json = serde_json::to_string_pretty(&data).unwrap();
    fs::write(filename, json).unwrap();
}

fn export_csv(names: &[String], filename: &str) {
    let mut csv = String::from("name\n");
    for n in names {
        csv.push_str(n);
        csv.push('\n');
    }
    fs::write(filename, csv).unwrap();
}

fn export_text(names: &[String], filename: &str) {
    fs::write(filename, names.join("\n")).unwrap();
}

fn main() {
    let matches = App::new("Historical Name Generator")
        .arg(Arg::with_name("culture").long("culture").takes_value(true).default_value("viking"))
        .arg(Arg::with_name("gender").long("gender").takes_value(true).default_value("male"))
        .arg(Arg::with_name("count").long("count").takes_value(true).default_value("1"))
        .arg(Arg::with_name("surname").long("surname"))
        .arg(Arg::with_name("title").long("title"))
        .arg(Arg::with_name("seed").long("seed").takes_value(true))
        .arg(Arg::with_name("output").long("output").takes_value(true))
        .arg(Arg::with_name("format").long("format").takes_value(true))
        .get_matches();

    let culture = matches.value_of("culture").unwrap();
    let gender = matches.value_of("gender").unwrap();
    let count: usize = matches.value_of("count").unwrap().parse().unwrap();
    let surname = matches.is_present("surname");
    let title = matches.is_present("title");
    let seed: u64 = matches.value_of("seed").unwrap_or("0").parse().unwrap();
    let output = matches.value_of("output");
    let format = matches.value_of("format");

    let mut gen = Generator::new(culture, gender, count, surname, title, seed);
    let names = gen.generate();

    if let Some(out) = output {
        let fmt = if let Some(f) = format { f } else {
            let ext = out.split('.').last().unwrap_or("txt");
            if ext == "json" { "json" } else if ext == "csv" { "csv" } else { "text" }
        };
        match fmt {
            "json" => export_json(&names, culture, gender, out),
            "csv" => export_csv(&names, out),
            _ => export_text(&names, out),
        }
        println!("Результат сохранён в {}", out);
    } else {
        let color = atty::is(atty::Stream::Stdout);
        print_names(&names, culture, gender, color);
    }
}
